File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.util.*;
import java.io.*;

class Main {
    static int n;
    static int m;
    static int[][] board;
    static int[][] visited;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    public static void bfs(int x, int y) {
        Deque<int[]> queue = new ArrayDeque<>();
        visited[x][y]++;
        queue.add(new int[]{x, y});

        while (!queue.isEmpty()) {
            int[] p = queue.poll();
            int px = p[0];
            int py = p[1];

            for (int i=0; i<4; i++) {
                int nx = px + dx[i];
                int ny = py + dy[i];

                if (nx >= 0 && nx < n && ny >= 0 && ny < m) {
                    if (visited[nx][ny] == 0 && board[nx][ny] == 1) {
                        visited[nx][ny] = visited[px][py] + 1;
                        queue.add(new int[]{nx, ny});
                    }
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        board = new int[n][m];
        visited = new int[n][m];

        for (int i=0; i<n; i++) {
            String s = br.readLine();
            for (int j=0; j<m; j++) {
                board[i][j] = s.charAt(j) - '0';
            }
        }

        bfs(0, 0);

        System.out.println(visited[n-1][m-1]);
    }
}

File: 백준/Silver/2178. 미로 탐색/미로 탐색.kt
fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine().split(" ").map { it.toInt() }
    val map = Array(n) { readLine().toCharArray().map { it - '0' }.toIntArray() }
    val visited = Array(n) { BooleanArray(m) }
    val dx = intArrayOf(-1, 1, 0, 0)
    val dy = intArrayOf(0, 0, -1, 1)
    
    val queue = ArrayDeque<Triple<Int, Int, Int>>() // (x, y, 거리)
    queue.add(Triple(0, 0, 1))
    visited[0][0] = true
    
    while (queue.isNotEmpty()) {
        val (x, y, dist) = queue.removeFirst()
        
        if (x == n - 1 && y == m - 1) {
            println(dist)
            return
        }
        
        for (d in 0..3) {
            val nx = x + dx[d]
            val ny = y + dy[d]
            
            if (nx in 0 until n && ny in 0 until m) {
                if (!visited[nx][ny] && map[nx][ny] == 1) {
                    queue.add(Triple(nx, ny, dist + 1))
                    visited[nx][ny] = true
                }
            }
        }
    }
}